# Gather paper notes

## Introduction

### Biological goals

Sensitively search and compare complex metagenomes.  Extract common components.

Identify known species and strains if present, or closest strains if not present.

Do so quickly and "inexpensively", for thousands to millions of samples.

Determine how much is not known.

Enable private databases.

(Taxonomic profiling vs taxonomic binning. Drawbacks of looking at independent reads vs k-mers/graphs.)

### MinHash/mash is awesome

Small and fixed-sized sketches. Data reduction.

Rapid comparison.

Robust and open and interoperable implementation.

Widely adopted approach (c.f. github repos, publications, etc.)

Linear search.

### New requirements => introduce sourmash

From the mash paper:

`Future applications of Mash could include read mapping and metagenomic sequence classification via windowed sketches or a containment score to test for the presence of one sequence within another [4]. However, both of these approaches would require additional sketch overhead to achieve acceptable sensitivity. Improvements in database construction are also expected. For example, rather than storing a single sketch per sequence (or window), similar sketches could be merged to further reduce space and improve search times. Obvious strategies include choosing a representative sketch per cluster or hierarchically merging sketches via a Bloom tree [40]. Finally, both the sketch and dist functions are designed as online algorithms, enabling, for example, dist to continually update a sketch from a streaming input. The program could then be modified to terminate when enough data have been collected to make a species identification at a predefined significance threshold. This functionality is designed to support the analysis of real-time data streams, as is expected from nanopore-based sequencing sensors [24].`


MinHash has poor sensitivity for really large "texts" (metagenomes).  This is acknowledged in the mash paper.

Containment is useful but fails due to resolution problem of MinHash approach when applied to samples of vastly different sizes, i.e. our problem.

Avoid linear search in mash: O(N). Conveniently SBTs just work for containment and you can truncate search very effectively.

(How big a hash do you need? --scaled solves this.)

## Methods

Mention k-mer abundance trimming approach from Zhang 2015 paper, and efficient implementation Zhang 201whatever paper.

## Results

### Some theory

"Banded shingling"

The math works out so that you can set a single resolution parameter (in addition to k) and build a collection of hashes in a single pass. This enables containment analysis.

Note MinHash compatibility within limits.

There may be some interesting stuff with errors and the width of the banding.

Math about missing windows of a given length Q with a scaled parameter of S; simulation => our software works.

(Maybe this is where we discuss k-mers and De Bruijn graphs more?)

### Measure SBT performance

Threshold of similarity vs speed (linear vs SBT approach) graph from SSBT paper.

Measure truncation of search rates.

Where do we classify a single genome with gather? Maybe here.

### Gather description and demonstration

Greedy search / optimal containment.

### Basic evaluation

1. Podar reads against 51 genomes that are actually in podar. (Comparison with genomes contained in specific reads.)
 => we get back what we know is in there. => gather works.
 
2. Podar reads against genbank. => we get back what we know is in there (including sherine stuff); => gather works fast; measure of false positive (nothing outside of genus)

### Comparison with kaiju/etc performance

http://ccb.jhu.edu/software/kraken/MANUAL.html#output-format

Kaiju (database -e) on Podar data vs Podar sourmash

Implement CAMI metrics (unifrac error, recall, l1norm error, precision, false positives)

Maybe compare with eukaryotes/how kaiju classifies eukaryotes?

## Discussion

### It basically works

Identifies down to a strain level which known genomes to use for further analysis.

Note on MetaPalette / k-mer size / strain / species tradeoff; rapid identification of strain variation.

Highly dependent on your reference data set.

Note no guarantee of contiguity in results.

### This is practical and convenient on laptops

Value of online data structures and algorithms in computation.

Updating / combining databases, building private databases, is all cheap.

Very tractable for private data sets.

Compatible with streaming (but not shown here).

### Tradeoff with hashset size is not bad in practice

The sampling factor (S) tells you how good the compression is. Bounded by constant factor times size of file. Do some quick measurements.

Here we can talk about unknown hashes; containment analysis being more useful for metagenomes.

### Future prospects / use cases

Because of the low false positive rate, this may be useful for detection of known contaminants.

You can identify known genomes in metagenomes very quickly.

# Additional material


### query times figure from SSBT paper
![](https://i.imgur.com/gPJgOxv.png)

### kaiju output example

```
1. either C or U, indicating whether the read is classified or unclassified.
2. name of the read
3. NCBI taxon identifier of the assigned taxon
4. the length or score of the best match used for classification
5. the taxon identifiers of all database sequences with the best match
6. the accession numbers of all database sequences with the best match
7. matching fragment sequence(s)

==> kaijudb_e_SRR606249.pe.qc.fq.gz.abundtrim.out <==
C	SRR606249.32811	131567	33	114,195,210,384,386,485,487,562,595,623,670,955,1280,1773,5061,5062,5454,5551,5854,5861,27334,28901,34388,36050,40296,44415,46354,51314,54262,57270,85471,86049,113226,121759,149539,204042,236234,254056,273371,288965,321614,332648,332952,334819,367110,426418,431241,431895,454130,510951,510952,572307,573508,644358,645133,647221,655827,655844,665079,756982,759272,759273,856822,864142,933388,985895,999810,1043628,1072389,1081102,1081103,1081104,1081105,1081107,1081108,1081109,1143189,1144522,1159556,1170229,1172203,1208600,1209926,1213859,1245745,1247875,1263415,1276135,1316194,1367422,1379787,1432657,1434259,1437433,1710895,1752066,1896206,1906856,	AQT24107.1,CCD49770.1,CDJ46319.1,CDJ53802.1,CDO63795.1,CDU16128.1,CDU16584.1,CDU16658.1,CDU16680.1,CDU16711.1,CDU17823.1,CDU17858.1,CDU18219.1,CDU19159.1,CDU19474.1,CDU19550.1,CDU19986.1,CDU20043.1,CDU20279.1,CDU20602.1,CDU20699.1,CDU84965.1,CDU85053.1,CEL07444.1,CEL65710.1,CEN61688.1,CRG99008.1,CUX78913.1,EGZ69942.1,EGZ70893.1,EGZ71490.1,EGZ71875.1,EGZ76067.1,EKV08921.1,EKV08938.1,EKV08943.1,EKV09315.1,EKV09659.1,ELQ33833.1,EPS26589.1,ETW57402.1,ETZ17808.1,EWC45822.1,GAQ39447.1,KDB11191.1,KDB11662.1,KDB11729.1,KDB11903.1,KDB12138.1,KDB12489.1,KDB12551.1,KDB13060.1,KDB13353.1,KDB14826.1,KDB16066.1,KDB16128.1,KDB16604.1,KDB17034.1,KDB17457.1,KDB17472.1,KDB17642.1,KDB17923.1,KDB18178.1,KFX91383.1,KGI40669.1,KGO44433.1,KGO67947.1,KGQ02786.1,KGQ03995.1,KGQ06756.1,KGQ10704.1,KHN98063.1,KHN98432.1,KHN98886.1,KHO00170.1,KID61264.1,KKZ58438.1,KKZ60663.1,KKZ62670.1,KKZ62773.1,KKZ67569.1,KLU85044.1,KLU92239.1,KMR48672.1,KMR67450.1,KMS34932.1,KNX47106.1,KNX49199.1,KPN27650.1,KUR53117.1,KXS96910.1,KYK91033.1,KZM21524.1,KZM23807.1,KZM24695.1,KZZ93757.1,KZZ94792.1,KZZ94976.1,KZZ95429.1,KZZ96569.1,OAA32658.1,OAA33279.1,OAA37109.1,OAA38176.1,OAA38178.1,OAA41149.1,OAA41528.1,OAA42811.1,OAA43505.1,OAA45878.1,OAA49302.1,OAA51317.1,OAA52510.1,OAA55819.1,OAA59882.1,OAA71978.1,OAA75345.1,OAA76529.1,OAG36567.1,OAL18638.1,OAL36509.1,OAL63226.1,OAL70576.1,OBQ24914.1,OBS19915.1,OCT54007.1,ODH46977.1,ODH50706.1,ODH51635.1,ODM18572.1,ODM20038.1,ODM21126.1,OHE95781.1,OHT03579.1,OKP10791.1,OKP14984.1,ONI56044.1,ONI56496.1,OOO11957.1,SBT32507.1,WP_010045825.1,WP_016942649.1,WP_031427251.1,WP_031737607.1,WP_033181154.1,WP_034329849.1,WP_041589251.1,WP_043582067.1,WP_048518638.1,WP_050893187.1,WP_052619396.1,WP_058039002.1,WP_064203791.1,WP_064203943.1,WP_065622016.1,WP_065980161.1,WP_065988210.1,WP_070329585.1,WP_071200734.1,WP_071526217.1,WP_071526219.1,WP_071526220.1,WP_071526221.1,WP_071526224.1,WP_071526246.1,WP_072276496.1,WP_073423533.1,WP_073983921.1,WP_074148240.1,WP_074148327.1,WP_074148330.1,WP_077166639.1,WP_077190508.1,WP_077190576.1,WP_077888963.1,XP_001556260.1,XP_001586711.1,XP_001748250.1,XP_001805962.1,XP_001940288.1,XP_002384145.1,XP_003839015.1,XP_003845926.1,XP_003870517.1,XP_003881788.1,XP_003882398.1,XP_003883670.1,XP_006696010.1,XP_006697324.1,XP_006961469.1,XP_006966882.1,XP_007285429.1,XP_007291787.1,XP_007297171.1,XP_007786797.1,XP_007800205.1,XP_007802225.1,XP_007806200.1,XP_007813799.1,XP_007820305.1,XP_008090405.1,XP_009851869.1,XP_009852655.1,XP_011124980.1,XP_013357929.1,XP_018159301.1,XP_018639329.1,XP_018642341.1,XP_018643437.1,XP_018693741.1,XP_018696364.1,XP_018707639.1,XP_018744148.1,XP_018749655.1,XP_019970036.1,XP_020132216.1,XP_957318.2,	KKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK,KKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK,KKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK,FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,
C	SRR606249.32811	131567	33	114,195,210,384,386,485,487,562,595,623,670,955,1280,1773,5061,5062,5454,5551,5854,5861,27334,28901,34388,36050,40296,44415,46354,51314,54262,57270,85471,86049,113226,121759,149539,204042,236234,254056,273371,288965,321614,332648,332952,334819,367110,426418,431241,431895,454130,510951,510952,572307,573508,644358,645133,647221,655827,655844,665079,756982,759272,759273,856822,864142,933388,985895,999810,1043628,1072389,1081102,1081103,1081104,1081105,1081107,1081108,1081109,1143189,1144522,1159556,1170229,1172203,1208600,1209926,1213859,1245745,1247875,1263415,1276135,1316194,1367422,1379787,1432657,1434259,1437433,1710895,1752066,1896206,1906856,	AQT24107.1,CCD49770.1,CDJ46319.1,CDJ53802.1,CDO63795.1,CDU16128.1,CDU16584.1,CDU16658.1,CDU16680.1,CDU16711.1,CDU17823.1,CDU17858.1,CDU18219.1,CDU19159.1,CDU19474.1,CDU19550.1,CDU19986.1,CDU20043.1,CDU20279.1,CDU20602.1,CDU20699.1,CDU84965.1,CDU85053.1,CEL07444.1,CEL65710.1,CEN61688.1,CRG99008.1,CUX78913.1,EGZ69942.1,EGZ70893.1,EGZ71490.1,EGZ71875.1,EGZ76067.1,EKV08921.1,EKV08938.1,EKV08943.1,EKV09315.1,EKV09659.1,ELQ33833.1,EPS26589.1,ETW57402.1,ETZ17808.1,EWC45822.1,GAQ39447.1,KDB11191.1,KDB11662.1,KDB11729.1,KDB11903.1,KDB12138.1,KDB12489.1,KDB12551.1,KDB13060.1,KDB13353.1,KDB14826.1,KDB16066.1,KDB16128.1,KDB16604.1,KDB17034.1,KDB17457.1,KDB17472.1,KDB17642.1,KDB17923.1,KDB18178.1,KFX91383.1,KGI40669.1,KGO44433.1,KGO67947.1,KGQ02786.1,KGQ03995.1,KGQ06756.1,KGQ10704.1,KHN98063.1,KHN98432.1,KHN98886.1,KHO00170.1,KID61264.1,KKZ58438.1,KKZ60663.1,KKZ62670.1,KKZ62773.1,KKZ67569.1,KLU85044.1,KLU92239.1,KMR48672.1,KMR67450.1,KMS34932.1,KNX47106.1,KNX49199.1,KPN27650.1,KUR53117.1,KXS96910.1,KYK91033.1,KZM21524.1,KZM23807.1,KZM24695.1,KZZ93757.1,KZZ94792.1,KZZ94976.1,KZZ95429.1,KZZ96569.1,OAA32658.1,OAA33279.1,OAA37109.1,OAA38176.1,OAA38178.1,OAA41149.1,OAA41528.1,OAA42811.1,OAA43505.1,OAA45878.1,OAA49302.1,OAA51317.1,OAA52510.1,OAA55819.1,OAA59882.1,OAA71978.1,OAA75345.1,OAA76529.1,OAG36567.1,OAL18638.1,OAL36509.1,OAL63226.1,OAL70576.1,OBQ24914.1,OBS19915.1,OCT54007.1,ODH46977.1,ODH50706.1,ODH51635.1,ODM18572.1,ODM20038.1,ODM21126.1,OHE95781.1,OHT03579.1,OKP10791.1,OKP14984.1,ONI56044.1,ONI56496.1,OOO11957.1,SBT32507.1,WP_010045825.1,WP_016942649.1,WP_031427251.1,WP_031737607.1,WP_033181154.1,WP_034329849.1,WP_041589251.1,WP_043582067.1,WP_048518638.1,WP_050893187.1,WP_052619396.1,WP_058039002.1,WP_064203791.1,WP_064203943.1,WP_065622016.1,WP_065980161.1,WP_065988210.1,WP_070329585.1,WP_071200734.1,WP_071526217.1,WP_071526219.1,WP_071526220.1,WP_071526221.1,WP_071526224.1,WP_071526246.1,WP_072276496.1,WP_073423533.1,WP_073983921.1,WP_074148240.1,WP_074148327.1,WP_074148330.1,WP_077166639.1,WP_077190508.1,WP_077190576.1,WP_077888963.1,XP_001556260.1,XP_001586711.1,XP_001748250.1,XP_001805962.1,XP_001940288.1,XP_002384145.1,XP_003839015.1,XP_003845926.1,XP_003870517.1,XP_003881788.1,XP_003882398.1,XP_003883670.1,XP_006696010.1,XP_006697324.1,XP_006961469.1,XP_006966882.1,XP_007285429.1,XP_007291787.1,XP_007297171.1,XP_007786797.1,XP_007800205.1,XP_007802225.1,XP_007806200.1,XP_007813799.1,XP_007820305.1,XP_008090405.1,XP_009851869.1,XP_009852655.1,XP_011124980.1,XP_013357929.1,XP_018159301.1,XP_018639329.1,XP_018642341.1,XP_018643437.1,XP_018693741.1,XP_018696364.1,XP_018707639.1,XP_018744148.1,XP_018749655.1,XP_019970036.1,XP_020132216.1,XP_957318.2,	KKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK,KKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK,KKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK,FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,
C	SRR606249.109181	1352	12	1352,	WP_073470582.1,WP_073470609.1,WP_073470624.1,	FFSSRRRHTRLV,
U	SRR606249.109181	0
C	SRR606249.130923	131567	33	114,195,197,210,384,386,485,487,562,595,623,670,955,1280,1773,5061,5062,5454,5551,5854,5861,27334,28901,34388,36050,40296,44415,46354,51314,54262,57270,85471,86049,113226,121759,149539,204042,236234,254056,273371,288965,321614,332648,332952,334819,367110,426418,431241,431895,454130,510951,510952,572307,573508,644358,645133,647221,655827,655844,665079,756982,759272,759273,856822,864142,933388,985895,999810,1036723,1043628,1072389,1081102,1081103,1081104,1081105,1081107,1081108,1081109,1143189,1144522,1159556,1170229,1172203,1208600,1209926,1213859,1245745,1247875,1263415,1276135,1316194,1367422,1379787,1432657,1434259,1437433,1710895,1752066,1896206,1906856,	AQT24107.1,CCD49770.1,CDJ46319.1,CDJ53802.1,CDO63795.1,CDU16128.1,CDU16584.1,CDU16658.1,CDU16680.1,CDU16711.1,CDU17823.1,CDU17858.1,CDU18219.1,CDU19159.1,CDU19474.1,CDU19550.1,CDU19986.1,CDU20043.1,CDU20279.1,CDU20602.1,CDU20699.1,CDU84965.1,CDU85053.1,CEL07444.1,CEL65710.1,CEN61688.1,CRG99008.1,CUX78913.1,EGZ69942.1,EGZ70893.1,EGZ71490.1,EGZ71875.1,EGZ76067.1,EKV08921.1,EKV08938.1,EKV08943.1,EKV09315.1,EKV09659.1,ELQ33833.1,EPS26589.1,ETW15539.1,ETW57402.1,ETZ17808.1,EWC45822.1,GAQ39447.1,KDB11191.1,KDB11662.1,KDB11729.1,KDB11903.1,KDB12138.1,KDB12489.1,KDB12551.1,KDB13060.1,KDB13353.1,KDB14826.1,KDB16066.1,KDB16128.1,KDB16604.1,KDB17034.1,KDB17457.1,KDB17472.1,KDB17642.1,KDB17923.1,KDB18178.1,KFX91383.1,KGI40669.1,KGO44433.1,KGO67947.1,KGQ02786.1,KGQ03995.1,KGQ06756.1,KGQ10704.1,KHN98063.1,KHN98432.1,KHN98886.1,KHO00170.1,KID61264.1,KKZ58438.1,KKZ60663.1,KKZ62670.1,KKZ62773.1,KKZ67569.1,KLU85044.1,KLU92239.1,KMR48672.1,KMR67450.1,KMS34932.1,KNX47106.1,KNX49199.1,KPN27650.1,KUR53117.1,KXS96910.1,KYK91033.1,KZM21524.1,KZM23807.1,KZM24695.1,KZZ93757.1,KZZ94792.1,KZZ94976.1,KZZ95429.1,KZZ96569.1,OAA32658.1,OAA33279.1,OAA37109.1,OAA38176.1,OAA38178.1,OAA41149.1,OAA41528.1,OAA42811.1,OAA43505.1,OAA45878.1,OAA49302.1,OAA51317.1,OAA52510.1,OAA55819.1,OAA59882.1,OAA71978.1,OAA75345.1,OAA76529.1,OAG36567.1,OAL18638.1,OAL36509.1,OAL63226.1,OAL70576.1,OBQ24914.1,OBS19915.1,OCT54007.1,ODH46977.1,ODH50706.1,ODH51635.1,ODM18572.1,ODM20038.1,ODM21126.1,OHE95781.1,OHT03579.1,OIT95932.1,OKP10791.1,OKP14984.1,ONI56044.1,ONI56496.1,OOO11957.1,SBT32507.1,WP_010045825.1,WP_016942649.1,WP_031427251.1,WP_031737607.1,WP_033181154.1,WP_034329849.1,WP_041589251.1,WP_043582067.1,WP_048518638.1,WP_050893187.1,WP_052619396.1,WP_058039002.1,WP_064203791.1,WP_064203943.1,WP_065622016.1,WP_065980161.1,WP_065988210.1,WP_070329585.1,WP_071200734.1,WP_071526217.1,WP_071526219.1,WP_071526220.1,WP_071526221.1,WP_071526224.1,WP_071526246.1,WP_072276496.1,WP_073423533.1,WP_073983921.1,WP_074148240.1,WP_074148327.1,WP_074148330.1,WP_077166639.1,WP_077190508.1,WP_077190576.1,WP_077888963.1,XP_001556260.1,XP_001586711.1,XP_001748250.1,XP_001805962.1,XP_001940288.1,XP_002384145.1,XP_003839015.1,XP_003845926.1,XP_003870517.1,XP_003881788.1,XP_003882398.1,XP_003883670.1,XP_006696010.1,XP_006697324.1,XP_006961469.1,XP_006966882.1,XP_007285429.1,XP_007291787.1,XP_007297171.1,XP_007786797.1,XP_007800205.1,XP_007802225.1,XP_007806200.1,XP_007813799.1,XP_007820305.1,XP_008090405.1,XP_009851869.1,XP_009852655.1,XP_011124980.1,XP_013357929.1,XP_018159301.1,XP_018639329.1,XP_018642341.1,XP_018643437.1,XP_018693741.1,XP_018696364.1,XP_018707639.1,XP_018744148.1,XP_018749655.1,XP_019970036.1,XP_020132216.1,XP_957318.2,	KKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK,KKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKN,KKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKT,FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,VFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,CFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,
C	SRR606249.130923	131567	25	210,623,1280,1773,28901,51314,51316,1036727,1752066,1906856,	CDI76602.1,CDJ52616.1,ETW48723.1,KNX48506.1,WP_031737607.1,WP_052619396.1,WP_058039002.1,WP_070329585.1,WP_072276496.1,WP_074196436.1,WP_078122264.1,	KKKKKKKKKKKKKKKKKKKKTKKKK,KKKKKKKKKKKKKKKKKKKKQKKKK,FFFFCFFFFFFFFFFFFFFFFFFFF,FFFFLFFFFFFFFFFFFFFFFFFFF,
U	SRR606249.143734	0
U	SRR606249.143734	0
C	SRR606249.153201	2759	33	5861,44415,51314,86049,332648,502780,572307,655827,1036723,1076935,1081102,1081103,1081107,1081109,1159556,1316194,	CCX33834.1,CDJ46319.1,CDU16658.1,ETW15539.1,KDB16604.1,KHN98063.1,KZZ94976.1,OAA37109.1,OAA45878.1,OAA59882.1,OCT54007.1,OKP10791.1,XP_001556260.1,XP_003882398.1,XP_007813799.1,XP_010763217.1,XP_013357929.1,	KEKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK,RKKRKKKKKKKKKKKKKKKKKKKKKKKKKKKKK,
C	SRR606249.153201	131567	29	1280,5861,	CDU19261.1,KMS30424.1,KMS34932.1,	KKKKKKKKKKKKKKKKKKKKKKKKKKIKK,

==> kaijudb_n_SRR606249.pe.qc.fq.gz.abundtrim.out <==
C	SRR606249.32811	131567
C	SRR606249.32811	131567
C	SRR606249.109181	1352
U	SRR606249.109181	0
C	SRR606249.130923	131567
C	SRR606249.130923	2
U	SRR606249.143734	0
U	SRR606249.143734	0
C	SRR606249.153201	1783272
C	SRR606249.153201	1783272
```
```
==> kaijudb_e_SRR606249.pe.qc.fq.gz.abundtrim.out.species.summary <==
        %	    reads	species
-------------------------------------------
11.789155	  1542113	Rhodopirellula baltica
 4.733213	   619141	Nitrosomonas europaea
 4.326456	   565934	Archaeoglobus fulgidus
 4.153454	   543304	Gemmatimonas aurantiaca
 3.816562	   499236	Nanoarchaeum equitans
 3.606063	   471701	Pyrococcus horikoshii
 3.248110	   424878	Sulfolobus tokodaii
 2.691461	   352064	Acidobacterium capsulatum
```

Rules for CAMI output:
https://github.com/CAMI-challenge/contest_information/blob/master/file_formats/CAMI_TP_specification.mkd